src/compute/src/arrangement/merge_batcher_col.rs

```diff
@@ -0,0 +1,491 @@
+                            keep = self.sorter.empty();
+                        }
+                    } else if keep.len() == keep.capacity() {
+                        kept.push(keep);
+                        keep = self.sorter.empty();
+                    }
+                    keep.copy(datum);
+                } else {
+                    builder.copy(datum);
+                }
+            }
+            // Recycling buffer.
+            self.sorter.recycle(buffer);
+        }
+
+        // Finish the kept data.
+        if !keep.is_empty() {
+            kept.push(keep);
+        }
+        if !kept.is_empty() {
+            self.sorter.push_list(kept);
+        }
+
+        // Drain buffers (fast reclamation).
+        self.sorter.clear_stash();
+
+        let seal = builder.done(
+            self.lower.clone(),
+            upper.clone(),
+            Antichain::from_elem(T::minimum()),
+        );
+        self.lower = upper;
+        seal
+    }
+
+    /// The frontier of elements remaining after the most recent call to `self.seal`.
+    fn frontier(&mut self) -> timely::progress::frontier::AntichainRef<T> {
+        self.frontier.borrow()
+    }
+}
+
+struct FixedStackQueue<T: Columnation> {
+    list: FixedStack<T>,
+    head: usize,
+}
+
+impl<T: Columnation> Default for FixedStackQueue<T> {
+    fn default() -> Self {
+        Self::from(Default::default())
+    }
+}
+
+impl<T: Columnation> FixedStackQueue<T> {
+    fn pop(&mut self) -> &T {
+        self.head += 1;
+        &self.list[self.head - 1]
+    }
+
+    fn peek(&self) -> &T {
+        &self.list[self.head]
+    }
+
+    fn from(list: FixedStack<T>) -> Self {
+        FixedStackQueue { list, head: 0 }
+    }
+
+    fn done(self) -> FixedStack<T> {
+        self.list
+    }
+
+    fn is_empty(&self) -> bool {
+        self.head == self.list[..].len()
+    }
+
+    /// Return an iterator over the remaining elements.
+    fn iter(&self) -> impl Iterator<Item = &T> + Clone + ExactSizeIterator {
+        self.list[self.head..].iter()
+    }
+}
+
+struct MergeSorterColumnation<
+    D: Columnation + 'static,
+    T: Columnation + 'static,
+    R: Columnation + 'static,
+> {
+    /// each power-of-two length list of allocations. Do not push/pop directly but use the corresponding functions.
+    queue: Vec<Vec<FixedStack<(D, T, R)>>>,
+    stash: Vec<FixedStack<(D, T, R)>>,
+    pending: Vec<(D, T, R)>,
+    logger: Option<Logger<DifferentialEvent, WorkerIdentifier>>,
+    operator_id: usize,
+}
+
+impl<
+        D: Ord + Columnation + 'static,
+        T: Ord + Columnation + 'static,
+        R: Semigroup + Columnation + 'static,
+    > MergeSorterColumnation<D, T, R>
+{
+    const BUFFER_SIZE_BYTES: usize = 64 << 10;
+
+    /// Buffer size (number of elements) to use for new/empty buffers.
+    const fn buffer_size() -> usize {
+        let size = std::mem::size_of::<(D, T, R)>();
+        if size == 0 {
+            Self::BUFFER_SIZE_BYTES
```

Comment:
uncovered

Reply:
(Moved to #25196) Expected; only used for zero-sized types which we don't construct.